exonum/src/runtime/dispatcher/schema.rs

```diff
@@ -14,194 +14,239 @@
-    pub(crate) fn artifacts(&self) -> MapIndex<T::Base, String, ArtifactSpec> {
-        self.access.clone().get_map(ARTIFACTS)
-    }
-
-    pub(super) fn pending_artifacts(&self) -> MapIndex<T::Base, String, ArtifactSpec> {
-        self.access.clone().get_map(PENDING_ARTIFACTS)
-    }
-
-    /// Set of launched service instances.
-    // TODO Get rid of data duplication in information schema. [ECR-3222]
-    pub(crate) fn service_instances(&self) -> MapIndex<T::Base, String, InstanceSpec> {
-        self.access.clone().get_map(SERVICE_INSTANCES)
-    }
-
-    /// Set of pending service instances.
-    // TODO Get rid of data duplication in information schema. [ECR-3222]
-    pub(super) fn pending_service_instances(&self) -> MapIndex<T::Base, String, InstanceSpec> {
-        self.access.clone().get_map(PENDING_INSTANCES)
-    }
-
-    /// Identifiers of launched service instances.
-    fn service_instance_ids(&self) -> MapIndex<T::Base, InstanceId, String> {
-        self.access.clone().get_map(INSTANCE_IDS)
-    }
-
-    /// Identifiers of pending service instances.
-    fn pending_instance_ids(&self) -> MapIndex<T::Base, InstanceId, String> {
-        self.access.clone().get_map(PENDING_INSTANCE_IDS)
+        Self {
+            artifacts: construct(&access, ARTIFACTS),
+            instances: construct(&access, INSTANCES),
+            instances_by_id: construct(&access, INSTANCE_IDS),
+            pending_artifacts: construct(&access, PENDING_ARTIFACTS),
+            pending_instances: construct(&access, PENDING_INSTANCES),
+        }
     }
 
     /// Returns the information about a service instance by its identifier.
     pub fn get_instance<'q>(
         &self,
         query: impl Into<InstanceQuery<'q>>,
-    ) -> Option<(InstanceSpec, DeployStatus)> {
+    ) -> Option<(InstanceSpec, InstanceStatus)> {
         match query.into() {
-            InstanceQuery::Id(id) => {
-                if let Some(instance_name) = self.service_instance_ids().get(&id) {
-                    self.service_instances()
-                        .get(&instance_name)
-                        .map(|spec| (spec, DeployStatus::Active))
-                } else if let Some(instance_name) = self.pending_instance_ids().get(&id) {
-                    self.pending_service_instances()
-                        .get(&instance_name)
-                        .map(|spec| (spec, DeployStatus::Pending))
-                } else {
-                    None
-                }
-            }
-
-            InstanceQuery::Name(instance_name) => self
-                .service_instances()
-                .get(instance_name)
-                .map(|spec| (spec, DeployStatus::Active))
-                .or_else(|| {
-                    self.pending_service_instances()
-                        .get(instance_name)
-                        .map(|spec| (spec, DeployStatus::Pending))
-                }),
+            InstanceQuery::Id(id) => self
```

Comment:
Hm. IDs are primary entity for service interactions, but ID lookup costs more than name lookup. Maybe it worth inverting the things?

However, it's a topic for a different PR for sure.